sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/TimestampPolicyFactory.java

```diff
@@ -0,0 +1,206 @@
+
+import java.io.Serializable;
+import java.util.Optional;
+import org.apache.beam.sdk.io.kafka.TimestampPolicy.PartitionContext;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.kafka.common.TopicPartition;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/**
+ * An extendable factory to create a {@link TimestampPolicy} for each partition at runtime by
+ * KafkaIO reader. Subclasses implement {@link #createTimestampPolicy}, which is invoked by
+ * the the reader while starting or resuming from a checkpoint. Two commonly used policies are
+ * provided. See {@link #withLogAppendTime()} and {@link #withProcessingTime()}.
+ */
+public abstract class TimestampPolicyFactory<KeyT, ValueT> implements Serializable {
+
+  /**
+   * Creates a TimestampPolicy for a partition. This is invoked by the reader at the start or while
+   * resuming from previous checkpoint.
+   *
+   * @param tp The returned policy applies to records from this {@link TopicPartition}.
+   * @param previousWatermark The latest check-pointed watermark. This is set when the reader
+   *           is resuming from a checkpoint. This is a good value to return by implementations
+   *           of {@link TimestampPolicy#getWatermark(PartitionContext)} until a better watermark
+   *           can be established as more records are read.
+   * @return
+   */
+  public abstract TimestampPolicy<KeyT, ValueT> createTimestampPolicy(
+    TopicPartition tp, Optional<Instant> previousWatermark);
+
+  /**
+   * A {@link TimestampPolicy} that assigns processing time to each record.
+   * Specifically, this is the timestamp when the record becomes 'current' in the reader.
+   * The watermark aways advances to current time.
+   */
+  public static <K, V> TimestampPolicyFactory<K, V> withProcessingTime() {
+    return new TimestampPolicyFactory<K, V>() {
+      @Override
+      public TimestampPolicy<K, V>
+      createTimestampPolicy(TopicPartition tp, Optional<Instant> previousWatermark) {
+        return new ProcessingTimePolicy<>();
+      }
+    };
+  }
+
+  /**
+   * A {@link TimestampPolicy} that assigns Kafka's log append time (server side ingestion time)
+   * to each record. The watermark for each Kafka partition is the timestamp of the last record
+   * read. If a partition is idle, the watermark advances roughly to 'current time - 2 seconds'.
+   * See {@link KafkaIO.Read#withLogAppendTime()} for longer description.
+   */
+  public static <K, V> TimestampPolicyFactory<K, V> withLogAppendTime() {
+    //return (tp, previousWatermark) -> new LogAppendTimePolicy<>(previousWatermark);
+    return new TimestampPolicyFactory<K, V>() {
+      @Override
+      public TimestampPolicy<K, V>
+      createTimestampPolicy(TopicPartition tp, Optional<Instant> previousWatermark) {
+        return new LogAppendTimePolicy<>(previousWatermark);
+      }
+    };
+  }
+
+  /*
+   * TODO
```

Comment:
I like this function which is based on records in last period( 99% or min for example), will you add it?